File: 4_大津二值化/c++/OtsuThresholding.cpp

#include <iostream>
#include <opencv2/opencv.hpp>
 
using namespace std;
using namespace cv;
 
Mat RGB2Gray(const Mat& srcImage)
{
    int cols = srcImage.cols;
    int rows = srcImage.rows;
    Mat destImage = Mat::zeros(rows, cols, CV_8UC1);

    for (int i = 0; i < rows; ++i)
    {
        for (int j = 0; j < cols; ++j)
        {
            // 0.2126\ R + 0.7152\ G + 0.0722\ B
            destImage.ptr<unsigned char>(i)[j] = 0.2126 * srcImage.ptr<Vec3b>(i)[j][0] \
                + 0.7152 * srcImage.ptr<Vec3b>(i)[j][1] \
                + 0.0722 * srcImage.ptr<Vec3b>(i)[j][2];
        }
    }

    return destImage;
}

Mat Thresholding(const Mat& srcImage, int threshold = 128)
{
    int cols = srcImage.cols;
    int rows = srcImage.rows;

    Mat destImage = Mat::zeros(rows, cols, CV_8UC1);
    for (int i = 0; i < rows; ++i)
    {
        for (int j = 0; j < cols; ++j)
        {
            if (srcImage.ptr<unsigned char>(i)[j] >= 128)
            {
                destImage.ptr<unsigned char>(i)[j] = 255;
            }
            else
            {
                destImage.ptr<unsigned char>(i)[j] = 0;
            }
        }
    }

    return destImage;
}

int OtsuThresholdingValue(const Mat& srcImage)
{
    int th = 0;
    int max_sigma = 0;
    for (int i = 1; i <= 255; ++i)
    {
        int nCountUp = 0;
        int nCountDown = 0;
        int nSumUp = 0;
        int nSumDown = 0;
        for (int j = 0; j < srcImage.rows; ++j)
        {
            for (int k = 0; k < srcImage.cols; ++k)
            {
                int nValue = srcImage.ptr<unsigned char>(j)[k];
                if (nValue <= i)
                {
                    nCountDown++;
                    nSumDown += nValue;
                }
                else
                {
                    nCountUp++;
                    nSumUp += nValue;
                }
            }
        }

        float fMeanDown = nCountDown == 0 ? 0 : (float)nSumDown / (float)nCountDown;
        float fMeanUp = nCountUp == 0 ? 0 : (float)nSumUp / (float)nCountUp;
        float w0 = (float)nCountDown / (srcImage.cols * srcImage.rows);
        float w1 = (float)nCountUp / (srcImage.rows * srcImage.cols);

        float sigma = w0 * w1 * pow((fMeanDown - fMeanUp), 2);
        if (sigma > max_sigma)
        {
            max_sigma = sigma;
            th = i;
        }    
    }

    std::cout << "th: " << th << std::endl;
    return th;
}

int main()
{
    Mat srcImage = imread("../imori.jpg");
    imshow("Display Image window", srcImage);
    Mat destImage = RGB2Gray(srcImage);
    imshow("destImage", destImage);
    int th = OtsuThresholdingValue(destImage);
    destImage = Thresholding(destImage, th);
    imshow("threshold image", destImage);
    waitKey(0);
    return 0;
}

File: 4_大津二值化/python3/OtsuThresholding.py
import cv2
import numpy as np

def BGR2Gray(srcImage):

    b = srcImage[:,:,0]
    g = srcImage[:,:,1]
    r = srcImage[:,:,2]
    
    destImage = 0.2126 * r + 0.7152 * g + 0.0722 * b
    destImage = destImage.astype(np.uint8)

    return destImage

def OtsuThresholding(srcImage):
    max_sigma = 0
    max_t = 0
    
    for t in range(1, 255):
        v0 = srcImage[np.where(srcImage <= t)]
        if len(v0) > 0:
            m0 = np.mean(v0)
        else:
            m0 = 0
            
        w0 = len(v0) / (srcImage.shape[0] * srcImage.shape[1])
        
        v1 = srcImage[np.where(srcImage > t)]
        if (len(v1) > 0):
            m1 = np.mean(v1)
        else:
            m1 = 0
            
        w1 = len(v1) / (srcImage.shape[0] * srcImage.shape[1])
        
        sigma = w0 * w1 * ((m0 - m1)**2)
        if (sigma > max_sigma):
            max_sigma = sigma
            max_t = t
    
    print("threshold >>", max_t)
    return max_t
        
    

def Thresholding(srcImage, threshold = 128):
    srcImage[srcImage < threshold] = 0
    srcImage[srcImage > threshold] = 255
    return srcImage

srcImage = cv2.imread("imori.jpg")
cv2.imshow("imori", srcImage)

destImage = BGR2Gray(srcImage)
cv2.imshow("destImage", destImage)

max_t = OtsuThresholding(destImage)

destImage = Thresholding(destImage, max_t)
cv2.imshow("threshold image", destImage)

cv2.waitKey(0)
cv2.destroyAllWindows()